source/beamline/CLS/CLSSIS3820ScalerModeControl.cpp

```diff
@@ -0,0 +1,270 @@
+{
+	QString basicInfo = AMPseudoMotorControl::toString();
+
+	// Note this control's current value.
+
+	QString controlValue = "Value: " + toString(value());
+
+	// Note this control's tolerance.
+
+	QString controlTolerance = "Tolerance: " + QString::number(tolerance());
+
+	// Create and return complete info string.
+
+	QString result = basicInfo + "\n" + controlValue + "\n" + controlTolerance;
+
+	return result;
+}
+
+QString CLSSIS3820ScalerModeControl::toString(double value) const
+{
+	QString result = "";
+
+	switch (int(value)) {
+	case SingleShot:
+		result = "Single Shot";
+		break;
+	case Continuous:
+		result = "Continuous";
+		break;
+	case None:
+		result = "None";
+		break;
+	default:
+		result = "Unknown";
+		break;
+	}
+
+	return result;
+}
+
+void CLSSIS3820ScalerModeControl::setScanCountControl(AMControl *newControl)
+{
+	if (scanCountControl_ != newControl) {
+
+		if (scanCountControl_)
+			removeChildControl(scanCountControl_);
+
+		scanCountControl_ = newControl;
+
+		if (scanCountControl_)
+			addChildControl(scanCountControl_);
+
+		emit scanCountControlChanged(scanCountControl_);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::setNumberOfScansPerBufferControl(AMControl *newControl)
+{
+	if (numberOfScansPerBufferControl_ != newControl) {
+
+		if (numberOfScansPerBufferControl_)
+			removeChildControl(numberOfScansPerBufferControl_);
+
+		numberOfScansPerBufferControl_ = newControl;
+
+		if (numberOfScansPerBufferControl_)
+			addChildControl(numberOfScansPerBufferControl_);
+
+		emit numberOfScansPerBufferControlChanged(numberOfScansPerBufferControl_);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::setStartScanControl(AMControl *newControl)
+{
+	if (startScanControl_ != newControl) {
+
+		if (startScanControl_)
+			removeChildControl(startScanControl_);
+
+		startScanControl_ = newControl;
+
+		if (startScanControl_)
+			addChildControl(startScanControl_);
+
+		emit startScanControlChanged(startScanControl_);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::updateConnected()
+{
+	bool wasConnected = connected_;
+
+	bool isConnected = (
+				scanCountControl_ && scanCountControl_->isConnected() &&
+				numberOfScansPerBufferControl_ && numberOfScansPerBufferControl_->isConnected() &&
+				startScanControl_ && startScanControl_->isConnected()
+				);
+
+	setConnected(isConnected);
+
+	// If we weren't connected before but we are now, set mode to Continuous.
+
+	if (!wasConnected && isConnected)
+		move(Continuous);
+}
+
+void CLSSIS3820ScalerModeControl::updateValue()
+{
+	emit valueChanged(value_);
+}
+
+void CLSSIS3820ScalerModeControl::updateMoving()
+{
+	if (isConnected()) {
+		bool isMoving = (
+					scanCountControl_->isMoving() ||
+					numberOfScansPerBufferControl_->isMoving() ||
+					startScanControl_->isMoving()
+					);
+
+		setIsMoving(isMoving);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::onMoveSucceeded(QObject *action)
+{
+	AMPseudoMotorControl::onMoveSucceeded(action);
```

Comment:
Is the order of these two important?  Could there be issues of the control thinking it has succeeded before the setpoint is set?  Just curious.